packages/backend/src/corporate-proxy.ts

```diff
@@ -0,0 +1,103 @@
+ */
+export function configureCorporateProxyAgent() {
+  // Bootstrap global-agent, which addresses node-fetch proxy-ing.
+  // global-agent purposely uses namespaced env vars to prevent conflicting behavior with other libraries,
+  // but user can set GLOBAL_AGENT_ENVIRONMENT_VARIABLE_NAMESPACE to an empty value for global-agent to use
+  // the conventional HTTP_PROXY, HTTPS_PROXY and NO_PROXY environment variables.
+  // More details in https://github.com/gajus/global-agent#what-is-the-reason-global-agentbootstrap-does-not-use-http_proxy
+  bootstrap();
+
+  // Configure the undici package, which affects the native 'fetch'. It leverages the same env vars used by global-agent,
+  // or the more conventional HTTP(S)_PROXY ones.
+  const proxyEnv =
+    process.env.GLOBAL_AGENT_HTTP_PROXY ??
+    process.env.GLOBAL_AGENT_HTTPS_PROXY ??
+    process.env.HTTP_PROXY ??
+    process.env.http_proxy ??
+    process.env.HTTPS_PROXY ??
+    process.env.https_proxy;
+
+  if (proxyEnv) {
+    const proxyUrl = new URL(proxyEnv);
+
+    // Create an access token if the proxy requires authentication
+    let token: string | undefined = undefined;
+    if (proxyUrl.username && proxyUrl.password) {
+      const b64 = Buffer.from(
+        `${proxyUrl.username}:${proxyUrl.password}`,
+      ).toString('base64');
+      token = `Basic ${b64}`;
+    }
+
+    // Create a default agent that will be used for no_proxy origins
+    const defaultAgent = new Agent();
+
+    // Create an interceptor that will use the appropriate agent based on the origin and the no_proxy
+    // environment variable.
+    // Collect the list of domains that we should not use a proxy for.
+    // The only wildcard available is a single * character, which matches all hosts, and effectively disables the proxy.
+    const noProxyEnv =
+      process.env.GLOBAL_AGENT_NO_PROXY ??
+      process.env.NO_PROXY ??
+      process.env.no_proxy;
+    const noProxyList = noProxyEnv?.split(',') || [];
+
+    const isNoProxy = (origin: string | undefined): boolean => {
+      for (const exclusion of noProxyList) {
+        if (exclusion === '*') {
+          // Effectively disables proxying
+          return true;
+        }
+        // Matched as either a domain which contains the hostname, or the hostname itself.
+        if (origin === exclusion || origin?.endsWith(`.${exclusion}`)) {
+          return true;
+        }
+      }
+      return false;
+    };
+
+    const noProxyInterceptor = (
+      dispatch: Dispatcher['dispatch'],
+    ): Dispatcher['dispatch'] => {
+      return (opts, handler) => {
+        return isNoProxy(opts.origin?.toString())
+          ? defaultAgent.dispatch(opts, handler)
+          : dispatch(opts, handler);
+      };
+    };
+
+    // Create a proxy agent that will send all requests through the configured proxy, unless the
+    // noProxyInterceptor bypasses it.
+    const proxyAgent = new ProxyAgent({
```

Comment:
I don't think we can assume all HTTP calls will go through the proxyAgent.  For example, customers can customize the homepage and support page with their own links: https://github.com/janus-idp/backstage-showcase/blob/main/showcase-docs/customization.md.    If users click on those links in a proxied environment, they could be waiting a while for a response (300s seems to the be the default timeout).   We should document these areas where customers need to take some sort of action (remove links, update URL) and/or consider future support for shortening request timeouts

Reply:
Do the links that customers can add get loaded through RHDH? If they open in a new tab then it has nothing to do with the proxy settings, it would have to do with their workstation/network proxy settings I would think.

Reply:
Scaffolder is another area that may need to be considered

Reply:
> I don't think we can assume all HTTP calls will go through the proxyAgent. For example, customers can customize the homepage and support page with their own links: [main/showcase-docs/customization.md](https://github.com/janus-idp/backstage-showcase/blob/main/showcase-docs/customization.md?rgh-link-date=2024-05-13T18%3A07%3A45Z). If users click on those links in a proxied environment, they could be waiting a while for a response (300s seems to the be the default timeout). We should document these areas where customers need to take some sort of action (remove links, update URL) and/or consider future support for shortening request timeouts

These kinds of customized links will be accessed directly from the browser, no?
The settings here cover requests sent from the backend part of the application. As mentioned by @coreydaley above, client-side calls will typically be covered by browser/OS settings. I assume if their workstations are behind a corporate proxy, they already have instructions for configuring their browsers and systems accordingly.

Reply:
> Scaffolder is another area that may need to be considered

I just tested with some scaffolding templates (like https://github.com/redhat-developer/red-hat-developer-hub-software-templates/blob/main/templates/github/go-backend/template.yaml or https://github.com/backstage/software-templates/blob/main/scaffolder-templates/springboot-grpc-template/template.yaml), and could see the backend requests sent to the proxy, as expected.